components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt

```diff
@@ -192,12 +200,15 @@ class QrFragment : Fragment() {
             try {
                 image = reader.acquireNextImage()
                 val availableImage = image
-                if (availableImage != null) {
+                if (availableImage != null && scanCompleteListener != null) {
                     val source = readImageSource(availableImage)
                     val bitmap = BinaryBitmap(HybridBinarizer(source))
                     if (qrState == STATE_FIND_QRCODE) {
                         qrState = STATE_DECODE_PROGRESS
-                        AsyncScanningTask(scanCompleteListener).execute(bitmap)
+
+                        currentDecodeOperation = coroutineScope.launch {
```

Comment:
@Mugurell Since the scope is using `Dispatchers.Default`, it can have more than one active coroutine. If `tryScanningBitmap` is still running when the next `onImageAvailable` call comes in we'd be launching a second / third one, but we only keep track of the last one we launched (`currentDecodeOperation`). You can simulate this by adding a `delay` after `tryScanningBitmap`.

In `onStop` we would then only cancel the last one we launched but not the other ones that could still be running.

This doesn't really have any consequences as you set the `scanComplete` listener to null and other results that still come in would be ignored. But still this doesn't quite behave as intended?

I see the following options:
- Don't stop the jobs, let it finish and ignore results if the listener is null.
- Collect all jobs and cancel them all on stop.

**And these I wouldn't implement:**
- Limit to one coroutine/thread -- Wouldn't do that because that will make scanning slower unnecessarily
- Cancel previous coroutine -- Wouldn't do that either because we could lose a good result as the camera is moving

Reply:
Thank you for helping with this and sorry I hadn't responded earlier.
Postponed this a bit to think about the best way but forgot to update.

Based on you feedback I went with not trying to cancel coroutines but rather rely on null checks for `scanCompleteListener` for whether to continue execution or not.
Did not know of a small, clean way of collecting the coroutines and allow for them to automatically be disgarded when completed and if not for a small change this would've brought a bit too high of a complexity.
Also removed the `scanCompleteListener` as a parameter from `tryScanningBitmap` [since that would've leaked it](https://github.com/mozilla-mobile/android-components/blob/2a060eb2b9f80216690db1684cf0cfcaf4a83b5a/components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt#L617).

Would appreciate just a small pass over this changes, thanks again!

Reply:
@Mugurell OK, thanks, looks good! nit: we can remove the `currentDecodeOperation` field now, right, and also update the test name?

One more thing we could do defensively is set the `scanCompleteListener` to null once we have a successful scan. Since we're potentially doing that on multiple threads we could end up with multiple calls to `onScanComplete`. We could do in the `QrFeature` where we remove the fragment (similar to `onStop`). Wdyt?

Reply:
Indeed, I think there might be an edgecase here in which more Runnables about `onScanComplete` would be posted on Main.
.. Based on which we might again do inside `QrFeature#onScanComplete()`:
>            removeQrFragment()
>            onScanResult(result)

Went with nulling the `scanCompleteListener` in `QrFeature#onScanComplete()`.
Wanted to also add a test for this but.. seems like the invocation inside scanCompleteListener are not recorded.

Reply:
👍